### PR TITLE
Optimize peer storing in SQLite

### PR DESCRIPTION
**Pull request by braginini**
### Checklist
- [ ] Is it a bug fix
- [ ] Is a typo/documentation fix
- [x] Is a feature enhancement
- [ ] It is a refactor
- [ ] Created tests that fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

### CODE REVIEW

File: management/server/sqlite_store.go [hunk 1]
```diff
@@ -188,18 +211,38 @@ func (s *SqliteStore) SaveAccount(account *Account) error {
...
+			Create(account)
 		if result.Error != nil {
 			return result.Error
 		}
-		return nil
+
+		const batchSize = 500
+		err := batchInsert(account.PeersG, batchSize, tx)
+		if err != nil {
+			return err
```
Comment (lixmal) [inline management/server/sqlite_store.go:245]: Can we add context to all these, to help identify potential issues? We can probably deduce the location from the returned error already, but I'm not sure.
 
```suggestion
			return fmt.Errorf("insert PeersG: %w", err)
```

File: management/server/sqlite_store.go [hunk 2]
```diff
@@ -134,79 +135,159 @@ func (s *SqliteStore) AcquireAccountLock(accountID string) (unlock func()) {
...
+		key.AccountID = accCopy.Id
+		accCopy.SetupKeysG = append(accCopy.SetupKeysG, *key)
 	}
 
-	for id, peer := range account.Peers {
+	accCopy.PeersG = make([]nbpeer.Peer, 0, len(accCopy.Peers))
+	for id, peer := range accCopy.Peers {
 		peer.ID = id
-		account.PeersG = append(account.PeersG, *peer)
+		//we need an explicit reference to the account as it is missing for some reason
```
Comment (braginini) [inline management/server/sqlite_store.go]:
```suggestion
		//we need an explicit reference to the account for gorm
```

File: management/server/sqlite_store.go [hunk 3]
```diff
@@ -134,79 +135,159 @@ func (s *SqliteStore) AcquireAccountLock(accountID string) (unlock func()) {
...
+		peer.AccountID = accCopy.Id
+		accCopy.PeersG = append(accCopy.PeersG, *peer)
 	}
 
-	for id, user := range account.Users {
+	accCopy.UsersG = make([]User, 0, len(accCopy.Users))
+	for id, user := range accCopy.Users {
 		user.Id = id
+		//we need an explicit reference to an account as it is missing for some reason
```
Comment (braginini) [inline management/server/sqlite_store.go]:
```suggestion
		//we need an explicit reference to the account for gorm
```

File: management/server/sqlite_store.go (hunk 2 above)
Comment (braginini) [inline management/server/sqlite_store.go]:
```suggestion
		//we need an explicit reference to the account for gorm
```

Comment (braginini) [inline management/server/sqlite_store.go]:
```suggestion
		//we need an explicit reference to the account for gorm
```

Comment (braginini) [inline management/server/sqlite_store.go]:
```suggestion
		//we need an explicit reference to the account for gorm
```

File: management/server/sqlite_store.go [hunk 4]
```diff
@@ -134,79 +135,159 @@ func (s *SqliteStore) AcquireAccountLock(accountID string) (unlock func()) {
...
 func (s *SqliteStore) SaveAccount(account *Account) error {
 	start := time.Now()
 
-	for _, key := range account.SetupKeys {
-		account.SetupKeysG = append(account.SetupKeysG, *key)
+	// operate over a fresh copy as we will modify its fields
+	accCopy := account.Copy()
+	accCopy.SetupKeysG = make([]SetupKey, 0, len(accCopy.SetupKeys))
+	for _, key := range accCopy.SetupKeys {
+		//we need an explicit reference to the account as it is missing for some reason
```
Comment (braginini) [inline management/server/sqlite_store.go]:
```suggestion
		//we need an explicit reference to the account for gorm
```

File: management/server/sqlite_store.go [hunk 5]
```diff
@@ -134,79 +135,159 @@ func (s *SqliteStore) AcquireAccountLock(accountID string) (unlock func()) {
...
-		account.UsersG = append(account.UsersG, *user)
+		accCopy.UsersG = append(accCopy.UsersG, *user)
 	}
 
-	for id, group := range account.Groups {
+	accCopy.GroupsG = make([]nbgroup.Group, 0, len(accCopy.Groups))
+	for id, group := range accCopy.Groups {
 		group.ID = id
-		account.GroupsG = append(account.GroupsG, *group)
+		//we need an explicit reference to the account as it is missing for some reason
```
Comment (braginini) [inline management/server/sqlite_store.go]:
```suggestion
		//we need an explicit reference to the account for gorm
```

File: management/server/sqlite_store.go [hunk 6]
```diff
@@ -134,79 +135,159 @@ func (s *SqliteStore) AcquireAccountLock(accountID string) (unlock func()) {
...
+		group.AccountID = accCopy.Id
+		accCopy.GroupsG = append(accCopy.GroupsG, *group)
 	}
 
-	for id, route := range account.Routes {
+	accCopy.RoutesG = make([]route.Route, 0, len(accCopy.Routes))
+	for id, route := range accCopy.Routes {
 		route.ID = id
-		account.RoutesG = append(account.RoutesG, *route)
+		//we need an explicit reference to the account as it is missing for some reason
```
Comment (braginini) [inline management/server/sqlite_store.go]:
```suggestion
		//we need an explicit reference to the account for gorm
```

File: management/server/sqlite_store.go [hunk 7]
```diff
@@ -134,79 +135,159 @@ func (s *SqliteStore) AcquireAccountLock(accountID string) (unlock func()) {
...
+		route.AccountID = accCopy.Id
+		accCopy.RoutesG = append(accCopy.RoutesG, *route)
 	}
 
-	for id, ns := range account.NameServerGroups {
+	accCopy.NameServerGroupsG = make([]nbdns.NameServerGroup, 0, len(accCopy.NameServerGroups))
+	for id, ns := range accCopy.NameServerGroups {
 		ns.ID = id
-		account.NameServerGroupsG = append(account.NameServerGroupsG, *ns)
+		//we need an explicit reference to the account as it is missing for some reason
```
Comment (braginini) [inline management/server/sqlite_store.go]:
```suggestion
		//we need an explicit reference to the account for gorm
```

File: management/server/sqlite_store.go [hunk 8]
```diff
@@ -134,79 +135,159 @@ func (s *SqliteStore) AcquireAccountLock(accountID string) (unlock func()) {
 	return unlock
 }
 
+func batchInsert(records interface{}, batchSize int, tx *gorm.DB) error {
```
Comment (lixmal) [inline management/server/sqlite_store.go:138]: Why do we need this wrapper? Doesn't `tx.CreateInBatches` already do what we want?
If we still need the wrapper, this should ideally be a generic function for compile time type safety.

Comment (braginini) [reply management/server/sqlite_store.go:138]: How would you do it?
I don’t want to repeat the code for all objects

**Pull request closed without merging**